docs/prompting/self_criticism/self_refine.md

```diff
@@ -1,7 +1,135 @@
 ---
-title: ""
-description: ""
-keywords: ""
+description: "Self-Refine involves getting LLMs to iteratively generate new responses based off feedback or unitl a stopping condition is met"
 ---
 
-[wip]
+Self Refine <sup><a href="https://arxiv.org/pdf/2303.17651">1</a></sup> involves prompting a LLM to provide feedback on an answer. This iterative process continues until a stopping condition is met.
+
+We can implement this using `instructor` as seen below using our validation context below.
+
+```python hl_lines="25-27 56-65"
+import instructor
+from openai import OpenAI
+from pydantic import BaseModel, field_validator, ValidationInfo
+from typing import Literal
+
+client = instructor.from_openai(OpenAI())
+
+
+class Sentiment(BaseModel):
+    text: str
+
+    @field_validator("text")
+    @classmethod
+    def validate_text(cls, v: str, info: ValidationInfo) -> str:
+        if not v or not v.strip():
+            raise ValueError("Text must not be empty or whitespace")
+        pairwise_comparison_result = validate_sentiment(
+            v,
+            info.context["reference_statement"],  # type: ignore
+            info.context["sentiment"],  # type: ignore
+        )
+
+        if pairwise_comparison_result.alignment_result == "Review B":
+            raise ValueError(
+                f"""{pairwise_comparison_result.feedback}. Please modify
+                your statement to be more aligned with the target sentiment
+                and do not copy the statement provided for reference"""
+            )
+
+        if v == info.context["reference_statement"]:  # type: ignore
+            raise ValueError(
+                """Your statement is the same as the reference statement.
+                It should be a separate statement from the reference
+                statement."""
+            )
+
+        return v
+
+
+class PairwiseEvaluation(BaseModel):
+    feedback: str
+    alignment_result: Literal[
+        "Review A",
+        "Review B",
+        "Both",
+    ]
+
+
+def validate_sentiment(review_a: str, review_b: str, target_sentiment: str):
+    return client.chat.completions.create(
+        model="gpt-4o",
+        messages=[
+            {
+                "role": "system",
+                "content": f"""
+                Which review is aligned with the sentiment
+                {target_sentiment}?
+
+                Review A: {review_a}
+                Review B: {review_b}.
+
+                Pick your answer from ['Review A', 'Review B', 'both',
```

Comment:
should this part go into the docstring of `PairwiseEvaluation`

Reply:
Hmm I think it would be better in the prompt personally so that it's a bit clearer. Seems a bit abrupt to have a stand alone pick your answer in the docstring? Or are you referring to something like

```
class PairwiseEvaluation(BaseModel):
    """
  This is a class that represents a comparison of two reviews. Make sure to identify
   which one is more aligned with the target sentiment. Pick your answer from 
  the given alignment_results - Rview A, Review B or both
   """"
    feedback: str
    alignment_result: Literal[
        "Review A",
        "Review B",
        "Both",
    ]
```